### ISSUE TITLE
Unknown issues rendering images in Sine.html

### ISSUE THREAD

**Issue by sharun-s** [kiwix/kiwix-js#278]
Not sure what is happening here. Blob url is being created but seems to be empty
![math](https://user-images.githubusercontent.com/18413274/27897943-dc8329d6-6240-11e7-8eb6-e7dab5f75ec5.png)

**Comment by kelson42** [comment]
Hmmm, can the app display properly SVG files at all?

**Comment by sharun-s** [comment]
All the formula files are indeed svg files. Interestingly there are many files with extension  .svg.png which get displayed. Blobs with just .svg don't seem to display in these math formula related pages. Whether it's happening to all svg files...note sure yet. Have added a check to log svg loads.

The old evopedia code had some math formula rendering code. May or maynot be related to this issue.

**Comment by sharun-s** [comment]
This works. No svg/math formula issues. My mistake. I was testing the wrong directory with the wrong code. I am going to close this.

![math-2](https://user-images.githubusercontent.com/18413274/27917353-6639af18-6289-11e7-8913-31532b64a014.png)

**Comment by sharun-s** [comment]
I am going to reopen this. Sine.html article just doesn't load on any browser. All 3 freeze up.
[Issue heading can be changed to unknown issues with Sine.html]

Initially assumed it was some math formula rendering issue. Then assumed it was svg related.  Adding the following seemed to take care of it. But it was on my worker based image loader branch and that was only tested on Chrome. On recent retesting of the Sine.html page I find it never fully loaded. On all 3 browsers. Possibly related to #253. It seems to just gets stuck in some xzdec memory allocation loop.

```
if(util.endsWith(dirEntry.url.toLowerCase(), ".svg")){
    uiUtil.feedNodeWithBlob($(imgNodes[index]), 'src', content, 'image/svg+xml;');
}else{
    uiUtil.feedNodeWithBlob($(imgNodes[index]), 'src', content, 'image');
} 

```
I am seeing it fully load only on Chrome with my worker based image loader code. Firefox and Edge still  freeze up. It's not clear why this is making a difference yet. Reopening to track it.

**Comment by mossroy** [comment]
It looks like a different issue than the original one.
It would be better to create a different issue for that, else it makes it difficult to understand when you read it from the beginning.

**Comment by Jaifroid** [comment]
I've committed a patch to the Preload stylesheets branch that adds in `image/svg+xml` support in commit https://github.com/kiwix/kiwix-js/commit/030be01b3471c1e28f61958fb6d5a7da014d9961 . What I'm noticing is that I can "download" (inject) the formula images one-by-one in the Sine.html article in Edge/UWP using my click-on-image-to-show code, and they now show (with the patch), but if I try to download the images automatically, Edge freezes up. Looking at the console log in progressive download mode, these images are extracted very slowly. They seem to take much longer to pull out of the ZIM than simple binaries. Might it be something that is dealing with the xml? Effectively the progressive download code pulls 10 images fine (but slowly). On scroll, it tries to pull another 20 images, and bugs out on image 11.

The patch I've used to display image/svg+xml is this:
```
selectedArchive.readBinaryFile(dirEntry, function (readableTitle, content, namespace, url) {
    // TODO : use the complete MIME-type of the image (as read from the ZIM file)
    var mimetype = /^[^.]+\.svg$/i.test(url) ? "image/svg+xml" : "image";
    uiUtil.feedNodeWithBlob(image, 'src', content, mimetype);
```
NB, I previously patched readBinaryFile to return the namespace and the url along with the content, because I needed those for matching returned stylesheet blobs with their original entries, so that the stylesheets can be loaded in the correct order.

But note that TODO comment -- the person who wrote this suggests the mimetype can be extracted from the ZIM file. Pausing the code on one of these svg images reveals a mimetype of "6" stored in the dirEntry variable for the file. To decode this, we'd need to be able to read the [MimeTypeList](http://www.openzim.org/wiki/ZIM_file_format#MIME_Type_List_.28mimeListPos.29) contained in the ZIM file header. Presumably this varies for each ZIM file.

**Comment by Jaifroid** [comment]
OK, @sharun-s , I've discovered that if I set a sliceSize of 10 images both as initial slice and as subsequent slices, my progressive "download" branch can extract all the images on this page, but pretty slowly (on Edge/UWP):
![image](https://user-images.githubusercontent.com/4304337/28660737-19010768-72ac-11e7-8626-dbe0b18fade4.png)
This strongly suggests that the issue is to do with the directory lookup for svg files. Why? Because if I set an initial sliceSize of 10 and subsequent sliceSize of 20 (10 + 20), the program immediately seizes up on image 11, i.e. it copes with 10, pauses and waits for a scroll, and then hangs the moment it makes 20 image requests to getDirEntryByTitle, and doesn't extract even one image from that batch. If the second batch size is 10, no problem, they're all extracted but slowly and I can gradually go through the whole page. Again, I think we have a problem with asynchronous logjam on directory entry lookup, and it's not the blob extraction that's the problem, it's the lookup.

**Comment by Jaifroid** [comment]
By the way, @sharun-s , here's a better (fast) algorithm for setting the mimetype:
```
var mimetype = url.match(/\.(\w{2,4})$/);
mimetype = mimetype ? "image/" + mimetype[1].toLowerCase() : "image";
mimetype = /^[^\s]+\.jpg$/i.test(url) ? "image/jpeg" : mimetype;
mimetype = /^[^\s]+\.tif$/i.test(url) ? "image/tiff" : mimetype;
mimetype = /\.ico$/i.test(url) ? "image/x-icon" : mimetype;
mimetype = /^[^\s]+\.svg$/i.test(url) ? "image/svg+xml" : mimetype;
```

**Comment by Jaifroid** [comment]
I just want to document here that I've experimented with loading images without blobs to see if that might have been the issue. Instead of 
`uiUtil.feedNodeWithBlob(image, 'src', content, mimetype);`
It's possible to use:
```
var data = util.uintToString(content);
image[0].src = "data:" + mimetype + ";base64," + btoa(data);
```
This works fine, but is not faster than using blobs, and the directory entry lookup still causes the logjam above. It also clutters the page's source code with lots of base64 data.

**Comment by sharun-s** [comment]
@Jaifroid xzdec seems to get stuck requesting more memory. This is the stacktace I see when I hit the stop the script button (that browser pops up when things start freezing). 

Error: Script terminated by timeout at:
enlargeMemory@file:///K/www/js/lib/xzdec.js:1:18714
dynamicAlloc@file:///K/www/js/lib/xzdec.js:1:7557
_sbrk@file:///K/www/js/lib/xzdec.js:6:17095
_a@file:///K/www/js/lib/xzdec.js:10:34595
Ta@file:///K/www/js/lib/xzdec.js:10:8626
Xa@file:///K/www/js/lib/xzdec.js:10:20882
Ma@file:///K/www/js/lib/xzdec.js:10:2624
Decompressor.prototype._readLoop/<@file:///K/www/js/lib/xzdec_wrapper.js:75:23
_fulfilled@file:///K/www/js/lib/q.js:794:54
Promise.prototype.then/</<@file:///K/www/js/lib/q.js:823:30
Promise/promise.promiseDispatch@file:///K/www/js/lib/q.js:756:13
become/</<@file:///K/www/js/lib/q.js:564:17
flush@file:///K/www/js/lib/q.js:110:17

**Comment by Jaifroid** [comment]
@sharun-s , @mossroy , @kelson42 (in case he has insights on the ZIM format or experience from the other builds): We need to theorize what is going on here (summary: svg files require far more memory just to do a directory lookup on them in the ZIM than other image files). This is the exact same performance issue we saw with stylesheets. Why is it the case that binary images (jpg, png, bmp, etc.) are found extremely easily by the getDirEntryByTitle() function whereas stylesheets (basically ASCII) and svg files (which have xml data in them) seem to cause some kind of seeking behaviour that demands a lot of resources from the pseudo-binary xzdec.js?

The only thing I can think of that these formats (css and svg) have in common is the xml-like structure of the data. Might this trigger a search through the ZIM file for a title, for example, which neither file format has? Or have these been compressed differently (I guess jpeg is already a compressed format, so maybe stored in the ZIM differently)? For whatever reason, xzdec.js is treating these files differently from jpegs and other binary formats. Any thoughts/insights?

**Comment by sharun-s** [comment]
Root cause is xzdec is a blackbox.  The interface and how xzdec itself behaves (for example when multiple decompressions are happening in parallel, configurable options etc) needs to be documented. 

And that is one of those thankless laborious tasks where you have to put in the time to become a decompression expert to, at the end of the day, probably fix 1 line of code somewhere. 
Ideally we need a decompression guy to look at this/document the tweakable variables, constraints etc.

Until then workarounds will be the fallback.
The whole page loads quite fast in chrome with workers. Will have to put in some time to see why Firefox is choking.

**Comment by Jaifroid** [comment]
Combination of your workers with gradual download of the images (my slices) should do the trick. Firefox is pretty happy with Sine.html with the code I've implemented for svg's. Try it out with the latest version of the Preload Stylesheets branch -- currently commit https://github.com/kiwix/kiwix-js/commit/5fd129a77fa151e128858b674612707a9794f4bd -- and you'll see that Firefox sails through the page (watch the console.log) if the images are requested in discrete batches.

I added [this routine](https://github.com/kiwix/kiwix-js/blob/5fd129a77fa151e128858b674612707a9794f4bd/www/js/app.js#L1079) to control the batches if SVGs are detected in the slice stream.

The problem is really on mobile, where even controlling the slice sizes SVGs are very taxing to the system, as are stylesheets. It's entirely for that reason that I'm caching the stylesheets in the system. It's all very bizarre.

**Comment by sharun-s** [comment]
Yup it should. Adding your "batch" mode stuff to workers is on my todo list.

**Comment by Jaifroid** [comment]
Interestingly, the Spanish Wikipedia ZIM equivalent to the Sine article (Sinusoide.html) uses, or used, standard .png files as the fallback for mathematical formulae, and it loads as fast as any other page, despite showing many of the same images. This may have changed recently, because it looks like the online page uses svg fallbacks.

**Comment by Jaifroid** [comment]
@sharun-s , I forgot to crossreference here #297: a MathJax solution for the display of mathematical SVGs in Kiwix JS. It allows all the platforms I test on to render even the most complex pages full of maths formulae, such as "List of integrals of trigonometric functions" in English Wikipedia.

**Comment by kelson42** [comment]
@mossroy @Jaifroid I really believe this ticket can be closed too, I best case this is a duplicate of already open tickets.

**Comment by Jaifroid** [comment]
I agree.
